pkg/services/annotations/annotationsimpl/composite_store.go

```diff
@@ -0,0 +1,75 @@
+package annotationsimpl
+
+import (
+	"context"
+	"sort"
+
+	"github.com/grafana/grafana/pkg/services/annotations"
+	"github.com/grafana/grafana/pkg/services/annotations/accesscontrol"
+	"golang.org/x/sync/errgroup"
+)
+
+// CompositeStore is a read store that combines two or more read stores, and queries all stores in parallel.
+type CompositeStore struct {
+	readers []readStore
+}
+
+func NewCompositeStore(readers ...readStore) *CompositeStore {
+	return &CompositeStore{
+		readers: readers,
+	}
+}
+
+// Get returns annotations from all stores, and combines the results.
+func (c *CompositeStore) Get(ctx context.Context, query *annotations.ItemQuery, accessResources *accesscontrol.AccessResources) ([]*annotations.ItemDTO, error) {
+	itemCh := make(chan []*annotations.ItemDTO, len(c.readers))
+
+	eg, ctx := errgroup.WithContext(ctx)
+	for _, r := range c.readers {
+		r := r
+		eg.Go(func() error {
+			items, err := r.Get(ctx, query, accessResources)
+			itemCh <- items
+			return err
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return make([]*annotations.ItemDTO, 0), err
+	}
+
+	close(itemCh)
+	res := make([]*annotations.ItemDTO, 0)
+	for items := range itemCh {
+		res = append(res, items...)
+	}
+	sort.Sort(annotations.SortedItems(res))
+
+	return res, nil
+}
+
+// GetTags returns tags from all stores, and combines the results.
+func (c *CompositeStore) GetTags(ctx context.Context, query *annotations.TagsQuery) (annotations.FindTagsResult, error) {
```

Comment:
The implementation here looks very much like the implementation of `Get()` - the parallelism works the same way. Can we extract it to something shared so that it's more DRY?

For instance, the parallelism doesn't need to know about the actual type being fetched, it can make a closure over the result channel, and the goroutine/error waiting stuff can use a shared function.

Reply:
I initially wrote this to used a shared function, but because the methods being called take in different arguments, there's a lot of closure boilerplate that needs to be set up to make it work. I'll put up a quick commit to illustrate what that would look like, but IMO it's not a great approach given those constraints.